src/c_api/c_api.cc

```diff
@@ -502,12 +507,278 @@ XGB_DLL int XGDMatrixCreateFromMat_omp(const bst_float* data,
       }
     }
   }
+  // restore omp state
+  omp_set_num_threads(nthread_orig);
 
   mat.info.num_nonzero = mat.row_data_.size();
   *out  = new std::shared_ptr<DMatrix>(DMatrix::Create(std::move(source)));
   API_END();
 }
 
+
+
```

Comment:
I think there's 3x the code here that you need. Try a function like:
```
float get_dt_value(void*col_ptr, wchar_t*type, bst_uint row){
  switch(type){
    Case int32 if ... return nan else return ...
  }
}
```

I don't think there needs to be any other switch statements or even a class.